ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs

```diff
@@ -104,16 +94,43 @@ mkMempool
      , ValidateEnvelope blk
      )
   => MempoolEnv m blk -> Mempool m blk TicketNo
-mkMempool env = Mempool
-    { tryAddTxs      = implTryAddTxs      env
-    , removeTxs      = implRemoveTxs      env
-    , syncWithLedger = implSyncWithLedger env
-    , getSnapshot    = implGetSnapshot    env
-    , getSnapshotFor = implGetSnapshotFor env
-    , getCapacity    = implGetCapacity    env
-    , getTxSize      = mpEnvTxSize        env
+mkMempool mpEnv = Mempool
+    { tryAddTxs      = \txs -> do
```

Comment:
This `tryAddTxs` has a change in semantics that I hadn't noticed before.

Here's the telling observation: each call to the old `tryAddTxs` could induce multiple STM transactions, such as one per transaction in the given list. This new implementation will only ever create a single STM transaction.

I haven't carefully considered why that's an important behavior to preserve, but I think it is. The old transactions would each just attempt to add one tx to the mempool. The new transactions each try to add all of the txs to the mempool, or the longest prefix that would fit. EG If we consider multiple peers trying to add to our mempool, the one tx per transaction seems likely to exhibit better fairness.

But like I said, I haven't considered that very carefully. Perhaps Thomas wrote something relevant in the Consensus Report we could double-check.

Reply:
This is indeed true. I re-implemented this operation to preserve the same execution flow as before, which should ensure fairness here.

This being said, I might still be missing some comments specifying why, but I think what I wrote should be enough.

TODO: test the fairness.

Reply:
As this was addressed in a follow-up issue #3127, I'm closing this comment.